### ISSUE TITLE
IsAssociativeContainer does not ensure map classs

### ISSUE THREAD

**Issue by qZhang88** [sewenew/tokenizer#1]
I used your code in my project. However compile indicate error on `container.emplace(key, val.get<typename T::mapped_type>());`

I think the reason is that IsAssociativeContainer does ensure type T has mapped_type

``` c++
template <typename T>
struct IsAssociativeContainer
    : std::integral_constant<bool,
        HasInsert<T>::value && !HasPushBack<T>::value> {};

}

    template <typename T, typename std::enable_if<detail::IsAssociativeContainer<T>::value, int>::type = 0>
    T get() const {
        T container;
        if (const auto *p = std::get_if<Object>(&_value)) {
            for (const auto &[key, val] : *p) {
                **container.emplace(key, val.get<typename T::mapped_type>());**
            }
        } else {
            throw Error("type mismatch");
        }

        return container;
    }

```

**Comment by sewenew** [comment]
It seems the code does not work with gcc. Anyway, you can fix it with the following change:

```
container.emplace(key, val.template get<typename T::mapped_type>());
```

I'll do the fix and update the code latter.

Regards

**Comment by qZhang88** [comment]
great, thanks it fixs.

If it's convenient, I have another question. Howto setup cmake file, so abseil-cpp and re2 could all be build together, by 

``` cmake
add_subdirectory(third_party/abseil-cpp)
add_subdirectory(third_party/re2)
```
now it has link problem

``` bash
/usr/bin/ld: CMakeFiles/tiktoken_test.dir/test/tiktoken_test.cpp.o: in function `tokenizer::Tiktoken::_build_special_token_regex(std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned long, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, unsigned long> > > const&)':
tiktoken_test.cpp:(.text._ZN9tokenizer8Tiktoken26_build_special_token_regexERKSt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmSt4hashIS7_ESt8equal_toIS7_ESaISt4pairIKS7_mEEE[_ZN9tokenizer8Tiktoken26_build_special_token_regexERKSt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmSt4hashIS7_ESt8equal_toIS7_ESaISt4pairIKS7_mEEE]+0xf1): undefined reference to `re2::RE2::QuoteMeta[abi:cxx11](absl::string_view)'
/usr/bin/ld: CMakeFiles/tiktoken_test.dir/test/tiktoken_test.cpp.o: in function `bool re2::RE2::FindAndConsume<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*>(absl::string_view*, re2::RE2 const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*&&)':
tiktoken_test.cpp:(.text._ZN3re23RE214FindAndConsumeIJPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEEbPN4absl11string_viewERKS0_DpOT_[_ZN3re23RE214FindAndConsumeIJPNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEEbPN4absl11string_viewERKS0_DpOT_]+0x57): undefined reference to `re2::RE2::FindAndConsumeN(absl::string_view*, re2::RE2 const&, re2::RE2::Arg const* const*, int)'
collect2: error: ld returned 1 exit status
```
many thanks

**Comment by sewenew** [comment]
You can install an older version of RE2, e.g. [version 2023-03-01](https://github.com/google/re2/releases/tag/2023-03-01) do not need abseil-cpp, which makes life easier :)

By the way, I've committed the change, and you can try the latest code on main branch.

Regards

**Comment by qZhang88** [comment]
Yes, indeed, now compile finished success. Thanks very much. Great repo, learned a lot.

Later I found out, compile error is caused by abseil-cpp, if abseil-cpp is installed by add_subdirectory, then ABSL_ENABLE_INSTALL is marked OFF. 

```
if(NOT CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
  option(ABSL_ENABLE_INSTALL "Enable install rule" OFF)
else()
  option(ABSL_ENABLE_INSTALL "Enable install rule" ON)
endif()
````

But re2 by default enable installation, and export re2Targets, which need abslTargets to be exported, which however is not installed, since ABSL_ENABLE_INSTALL is set OFF.

So I commented out install config in re2, since I don't need dynamic library. And then compile works well.
